Add `MemCx<'mcx>` for `List<'mcx, T>`

workingjubilee opened the pull request:
List requires reallocation in order to push to it with safe Rust, but it is defined as allocated via palloc, which has problems with lifetimes. In order to head off these problems, and begin work on related ones, introduce a type describing a MemoryContext with a lifetime, including scoped (therefore lifetime-bound) access to the current context.

workingjubilee committed:
Add skeletal MemCx for experiment

workingjubilee committed:
Refactor List to use MemCcx

workingjubilee committed:
s/NulalbleDatum/NullableDatum

workingjubilee committed:
Remember to add the temporary context closure fn

workingjubilee committed:
Refactor FnCall to experimental MemCx impl

workingjubilee committed:
Fix borrowck

workingjubilee committed:
Quick and dirty test fix

workingjubilee committed:
Fix Drain lifetime sig for pg13+

workingjubilee committed:
Fix borrowck for pg12

workingjubilee commented on pgrx/src/pg_sys.rs:
```diff
@@ -14,11 +15,13 @@ pub use pgrx_pg_sys::*;
-        .get((index - 1) as _)
-        .expect("rt_fetch used out-of-bounds")
-        .cast()
+    mem::current_context(|cx| {
+        crate::list::List::<*mut core::ffi::c_void>::downcast_ptr_in_memcx(range_table, cx)
+            .expect("rt_fetch used on non-ptr List")
+            .get((index - 1) as _)
+            .expect("rt_fetch used out-of-bounds")
+            .cast()
+    })
```
This diff suggests to me my approach is kinda incorrect, since I don't really "use" the context.

workingjubilee replied:
Yeah, this is probably going to be the most annoying part here. Maybe this would be better captured as another parameter to List, so non-allocating but owning uses don't have to worry about this.

workingjubilee commented on pgrx/src/mem.rs:
```diff
@@ -0,0 +1,39 @@
+    }
+}
+
+/// Acquire the current context and operate inside it.
+pub fn current_context<'curr, F, T>(f: F) -> T
+where
+    F: for<'clos> FnOnce(&'clos MemCx<'curr>) -> T,
+{
+    let memcx = unsafe { MemCx::from_ptr(pg_sys::CurrentMemoryContext) };
+    let ret = { f(&memcx) };
+    ret
+}
```
Definitely not sure about soundness, but something similar to this is probably required, for ergonomics reasons.

workingjubilee commented on pgrx/src/list/flat_list.rs:
```diff
@@ -357,25 +361,25 @@ unsafe fn destroy_list(list: *mut pg_sys::List) {
 /// A list being drained.
 #[derive(Debug)]
-pub struct Drain<'a, T> {
+pub struct Drain<'a, 'cx, T> {
```
Really need to move Drain itself "up".

Also: two lifetimes! spooky.

workingjubilee reviewed (commented):
UX thoughts: I probably didn't write the interfaces correctly. Might need redesign.

Wondered about maybe instead using something like `List<T, M>`, where `M` might be the thing infected with the lifetime? idk.

eeeebbbbrrrr commented:
Just a note to say I've skimmed over this.  I don't have any feedback yet other than it's exciting to see some code demonstrating the ideas.

workingjubilee commented on pgrx-tests/src/tests/list_tests.rs:
```diff
@@ -13,44 +13,44 @@ mod tests {
+            assert_eq!(70, list.len());
+            let _ = list.drain(50..);
+            assert_eq!(50, list.len());
+        })
```
Making sure Drain worked here was the most annoying thing to fix, since it wasn't intuitive until actually trialed, but was just an amount of lifetime poking so it was pretty okay, and our own implementation primitives are allowed to be messy to work with, I think: I don't expect anyone will hold on to a Drain for long.

workingjubilee commented on pgrx/src/fn_call.rs:
```diff
@@ -336,69 +337,79 @@ fn lookup_fn(fname: &str, args: &[&dyn FnCallArg]) -> Result<pg_sys::Oid> {
-            // if that didn't find a function, maybe we've got some defaults in there, so do a lookup
-            // where Postgres will consider that
-            fnoid = pg_sys::LookupFuncName(
+    mem::current_context(|mcx| {
```
The need to do this, and the problems with memory leaks in #1341, strongly suggests we should be accepting a MemCx argument to this function and also the public interface of `fn_call`.

thomcc reviewed (approved):
This is a good improvement.

workingjubilee committed:
Bikeshed pgrx::mem to pgrx::memcx

workingjubilee committed:
Add explaining comment

workingjubilee committed:
Expand docs on memory contexts

workingjubilee commented:
Rebased and added _some_ (~300 words) of documentation on memory contexts. I think I will land this now, so that it's easier to build off of and mutate as-we-go.

workingjubilee committed:
Aside about GCs

workingjubilee committed:
Fix tests using memcx

workingjubilee committed:
Fix pg12 again
